### PR TITLE
[MIG] business_requirement_deliverable: Migration to 11.0

### PR DESCRIPTION
**Pull request by martintorresvm**
Work in progress. Depends in :

- [x] https://github.com/OCA/business-requirement/pull/309

### CODE REVIEW

**Comment by martintorresvm** [comment]
cc @Tecnativa, @pedrobaeza @Tardo 

Pending for review and add reports

**Comment by martintorresvm** [comment]
I forced small changes. Problem here is no way of test on Runbot until base module is merged

File: business_requirement_deliverable/models/business.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+                   ('confirmed', 'Confirmed'),
+                   ('approved', 'Approved'),
+                   ('stakeholder_approval', 'Stakeholder Approval'),
+                   ('in_progress', 'In progress'),
+                   ('done', 'Done'),
+                   ('cancel', 'Cancel'),
+                   ('drop', 'Drop'),
+                   ],
+        store=True,
+    )
```
Comment (sergio-teruel) [inline business_requirement_deliverable/models/business.py:31]: This field is related without readonly, are you dure?
I do not remember if in v11 all related fields are readonly=true by default

File: business_requirement_deliverable/models/business.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    name = fields.Text('Name', required=True)
+    user_case = fields.Html()
+    proposed_solution = fields.Html()
+    product_id = fields.Many2one(
+        comodel_name='product.product',
+        string='Product',
+        domain=[('sale_ok', '=', True)],
+        required=False
```
Comment (sergio-teruel) [inline business_requirement_deliverable/models/business.py:39]: Remove this attribute, by default is false

File: business_requirement_deliverable/models/business.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    )
+    uom_id = fields.Many2one(
+        comodel_name='product.uom',
+        string='UoM',
+        required=True,
+        default=lambda self: self.env.ref('product.product_uom_unit')
+    )
```
Comment (sergio-teruel) [inline business_requirement_deliverable/models/business.py:46]: If you use a external id in default you need add an init hook and create the field manually

File: business_requirement_deliverable/models/business.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        string='Business Requirement',
+        ondelete='cascade',
+        required=True
+    )
+    sale_price_unit = fields.Float(
+        string='Sales Price',
+        oldname='unit_price'
+    )
+    price_total = fields.Float(
+        compute='_compute_get_price_total',
```
Comment (sergio-teruel) [inline business_requirement_deliverable/models/business.py]: As a convention, _compute_price_total

File: business_requirement_deliverable/models/business.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        store=True,
+        readonly=True
+    )
+    currency_id = fields.Many2one(
+        comodel_name='res.currency',
+        string='Currency',
+        readonly=True,
+        compute='_compute_get_currency',
+    )
+    business_requirement_partner_id = fields.Many2one(
```
Comment (sergio-teruel) [inline business_requirement_deliverable/models/business.py:81]: Why not use only partner_id

File: business_requirement_deliverable/models/business.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+        comodel_name='res.partner',
+        related='business_requirement_id.partner_id',
+        string='Stakeholder',
+        store=True
+    )
+    business_requirement_project_id = fields.Many2one(
```
Comment (sergio-teruel) [inline business_requirement_deliverable/models/business.py]: Project_id??

File: business_requirement_deliverable/models/business.py [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+from odoo.exceptions import UserError
+from odoo.addons import decimal_precision as dp
+
+
+class BusinessRequirementDeliverable(models.Model):
+    _name = "business.requirement.deliverable"
+    _inherit = ["mail.thread", "mail.activity.mixin"]
+    _description = "Business Requirement Deliverable"
+
+    sequence = fields.Integer('Sequence')
```
Comment (pedrobaeza) [inline business_requirement_deliverable/models/business.py:18]: Missing `_order = "sequence,id"`

File: business_requirement_deliverable/models/business.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+        selection=[('draft', 'Draft'),
+                   ('confirmed', 'Confirmed'),
+                   ('approved', 'Approved'),
+                   ('stakeholder_approval', 'Stakeholder Approval'),
+                   ('in_progress', 'In progress'),
+                   ('done', 'Done'),
+                   ('cancel', 'Cancel'),
+                   ('drop', 'Drop'),
+                   ],
+        store=True,
```
Comment (pedrobaeza) [inline business_requirement_deliverable/models/business.py:29]: Why is this needed? It requires `readonly=True`

Comment (martintorresvm) [reply business_requirement_deliverable/models/business.py:29]: Is to control others fields on views to be editable but you are right need to be readonly true

**Comment by martintorresvm** [comment]
Changes done

File: business_requirement_deliverable/models/business.py [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+    )
+    business_requirement_partner_id = fields.Many2one(
+        comodel_name='res.partner',
+        related='business_requirement_id.partner_id',
+        string='Stakeholder',
+        store=True
+    )
+    state = fields.Selection(related='business_requirement_id.state',
+                             string='State', store=True, readonly=True)
+    portal_published = fields.Boolean('In Portal', default=True)
```
Comment (sergio-teruel) [inline business_requirement_deliverable/models/business.py:89]: Why not use website_published and inherit from website.published.mixin

File: business_requirement_deliverable/models/business.py (hunk 1 above)
Comment (pedrobaeza) [reply business_requirement_deliverable/models/business.py:31]: In v11 default is `readonly=False`, so we have to explicitly put `readonly=True`

File: business_requirement_deliverable/models/business.py (hunk 3 above)
Comment (pedrobaeza) [reply business_requirement_deliverable/models/business.py:46]: Not this time, @sergio-teruel. That only happens when the external XML-ID is loaded in the same module. As this reference belongs to another module, there's no problem.

File: business_requirement_deliverable/models/business.py (hunk 9 above)
Comment (pedrobaeza) [reply business_requirement_deliverable/models/business.py:89]: Because with that, it will depend on `website` module, which is a heavy dependency.

File: business_requirement_deliverable/models/business.py (hunk 5 above)
Comment (pedrobaeza) [reply business_requirement_deliverable/models/business.py:81]: I suppose that this is because it's a related field.

File: business_requirement_deliverable/views/business_view.xml [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+        <field name="arch" type="xml">
+            <form string="Business Requirement Task Type">
+                <sheet>
+                    <div class="oe_button_box" name="button_box" groups="base.group_user">
+                         <button class="oe_inline oe_stat_button" name="portal_publish_button" type="object" icon="fa-globe">
+                             <label for="portal_published" />
+                             <field name="portal_published" widget="toggle_button"/>
+                         </button>
+                    </div>
+                    <group>
```
Comment (chienandalu) [inline business_requirement_deliverable/views/business_view.xml:121]: The title is going to better hanging from the `<sheet>`

**Comment by martintorresvm** [comment]
I pushed force dependency, monetary on Tree view, name search and group by section.

**Comment by martintorresvm** [comment]
Please review @sergio-teruel @pedrobaeza

File: business_requirement_deliverable/views/business_view.xml (hunk 10 above)
Comment (martintorresvm) [reply business_requirement_deliverable/views/business_view.xml:121]: Not sure how to do that, could you add suggested change?

**Comment by Tardo** [comment]
Added dependencies in portal page

File: business_requirement_deliverable/views/business_view.xml (hunk 10 above)
Comment (chienandalu) [reply business_requirement_deliverable/views/business_view.xml:121]:
```suggestion
```

File: business_requirement_deliverable/views/business_view.xml [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+                        <div class="oe_title">
+                            <h1>
+                                <field name="name" attrs="{'readonly':[('state','not in',['draft', 'confirmed'])]}"/>
+                            </h1>
+                        </div>
+                    </group>
```
Comment (chienandalu) [inline business_requirement_deliverable/views/business_view.xml:127]:
```suggestion
```

File: business_requirement_deliverable/views/business_view.xml [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+                <sheet>
+                    <div class="oe_button_box" name="button_box" groups="base.group_user">
+                         <button class="oe_inline oe_stat_button" name="portal_publish_button" type="object" icon="fa-globe">
+                             <label for="portal_published" />
+                             <field name="portal_published" widget="toggle_button"/>
+                         </button>
+                    </div>
+                    <group>
+                        <div class="oe_title">
+                            <h1>
```
Comment (chienandalu) [inline business_requirement_deliverable/views/business_view.xml:123]:
```suggestion
                            <h1 class="o_row">
```

File: business_requirement_deliverable/views/business_view.xml [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+                                <field name="name" attrs="{'readonly':[('state','not in',['draft', 'confirmed'])]}"/>
```
Comment (chienandalu) [inline business_requirement_deliverable/views/business_view.xml:124]:
```suggestion
                                <field name="name" attrs="{'readonly':[('state','not in',['draft', 'confirmed'])]}" placeholder="Deliverable title..."/>
```

File: business_requirement_deliverable/views/business_view.xml (hunk 10 above)
Comment (chienandalu) [reply business_requirement_deliverable/views/business_view.xml:121]: (it's not possible to add multiline suggestions, so you'll have to fix indent though)

**Comment by rafaelbn** [comment]
Depends in:

- [ ] https://github.com/OCA/business-requirement/pull/309

**Comment by martintorresvm** [comment]
Rebased

File: business_requirement_deliverable/views/business_view.xml (hunk 10 above)
Comment (martintorresvm) [reply business_requirement_deliverable/views/business_view.xml:121]: get it it's about the indent

**Review by rafaelbn — APPROVED** [review]
Using it already in production without issues :+1: 

Maybe some improvements are needed in the future
